internal/util/state/provider.go

```diff
@@ -22,88 +21,127 @@ import (
+	}
+
+	if p.filename != "" {
+		b, _ := os.ReadFile(p.filename)
+		_ = json.Unmarshal(b, p.s)
 	}
 
-	if _, err := p.Get(); err != nil {
-		return nil, err
+	p.s.fill()
+
+	// Simply overwrite state to handle all errors and edge cases
+	// like missing directory, corrupted file, invalid UUID, etc.
+	if err := persistState(p.s, p.filename); err != nil {
+		return p, fmt.Errorf("failed to persist state: %w", err)
 	}
 
 	return p, nil
 }
 
+// MetricsCollector return Prometheus metrics collector for that provider.
+//
 // If addUUIDToMetric is true, then the UUID is added to the Prometheus metric.
 func (p *Provider) MetricsCollector(addUUIDToMetric bool) prometheus.Collector {
 	return newMetricsCollector(p, addUUIDToMetric)
 }
 
-// Get returns the current process state.
+// Get returns a copy of the current process state.
 //
 // It is okay to call this function often.
 // The caller should not cache result; Provider does everything needed itself.
-func (p *Provider) Get() (*State, error) {
-	// return different copies to each caller
+func (p *Provider) Get() *State {
 	p.rw.RLock()
-	s := p.s
-	p.rw.RUnlock()
+	defer p.rw.RUnlock()
 
-	if s.UUID != "" {
-		return &s, nil
-	}
+	return p.s.deepCopy()
+}
 
-	b, _ := os.ReadFile(p.filename)
-	_ = json.Unmarshal(b, &s)
-	_, err := uuid.Parse(s.UUID)
+// Subscribe returns a channel that would receive notifications on state changes.
+// One notification would be scheduled immediately.
+func (p *Provider) Subscribe() chan struct{} {
+	p.rw.Lock()
+	defer p.rw.Unlock()
 
-	if err == nil {
-		// store a copy
-		p.rw.Lock()
-		p.s = s
-		p.rw.Unlock()
+	ch := make(chan struct{}, 1)
+	ch <- struct{}{}
 
-		return &s, nil
-	}
+	p.subs[ch] = struct{}{}
 
-	// all errors (missing file, invalid file permission, invalid JSON, etc)
-	// are handled in the same way - by regenerating state
+	return ch
+}
 
-	s.UUID = must.NotFail(uuid.NewRandom()).String()
-	b = must.NotFail(json.Marshal(s))
+// Unsubscribe removes the channel from subscriptions and closes it.
+func (p *Provider) Unsubscribe(ch chan struct{}) {
+	p.rw.Lock()
+	defer p.rw.Unlock()
 
-	if err := os.MkdirAll(filepath.Dir(p.filename), 0o777); err != nil && !os.IsExist(err) {
-		return nil, fmt.Errorf("failed to create state directory: %w", err)
+	delete(p.subs, ch)
+
+	close(ch)
+}
+
+// Update gets the current state, calls the given function, updates state, and notifies all subscribers.
+func (p *Provider) Update(update func(s *State)) error {
+	p.rw.Lock()
+	defer p.rw.Unlock()
+
+	update(p.s)
+	p.s = p.s.deepCopy()
+	p.s.fill()
+
+	err := persistState(p.s, p.filename)
+	if err != nil {
+		err = fmt.Errorf("failed to persist state: %w", err)
 	}
 
-	if err := os.WriteFile(p.filename, b, 0o666); err != nil {
-		return nil, fmt.Errorf("failed to write state file: %w", err)
+	// skip subscribers that already have notification waiting for them
+	for ch := range p.subs {
+		select {
+		case ch <- struct{}{}:
```

Comment:
This implementation looks safe to me (e.g. if we called `Unsubscribe` before, we don't have a chance to write into `ch` again as it will be out of `p.subs`). Still, I wonder if we need some tests, especially the ones that cover situations when `ch` is already full and that we don't wait for it.

Reply:
Done by removing `Unsubscribe` since we don't actually need it